### DISCUSSION TITLE
Dropping framerate in long campaign

### DISCUSSION THREAD

**Discussion by Shadowdash19** [FakeFishGames/Barotrauma#14716] [Bug reports]
### Disclaimers

- [X] I have searched the issue tracker to check if the issue has already been reported.
- [ ] My issue happened while using mods.

### What happened?

So i am kinda hoarder and when starting new campaign, we did arrive at 2nd location on boat already filled with resourses. So, what i did notice, our crew framerate really depends on items stashed on sub. At last we just ended playthrough, because framerate dropped so hard it was unplayable. So, yeah we DID play with mods, BUT after that i tested things out and after starting save without mods framerate was just slitly higher 20 avg FPS without mods against 17 avg FPS with mods. I thought what could be the issue? So i tried to start new campaign - framerate was just fine, even without performance fix. I did start campaign with lot of bots in crew - and still, fps was good. So i would guess, that trouble is game poorly can handle processing of large amount of items and lag out. I decided that must some kind of bug or misoptimization, because other sandbox games - like minecraft, for example, can handle big stash just fine. So i'm not into programming or that kind of things so i can only wonder what could be the issue. I just really want game not to punish me with unplayable performance for my playstyle. Maybe you could pay a little bit of attention to that trouble, please?

### Reproduction steps

1. Start any kind of lobby with preferably big sub
2. Spawn a lot of items in all containers 
3. Enjoy framerate

### Bug prevalence

Happens every time I play

### Single player or multiplayer?

Multiplayer hosted from the in-game menu (= using a listen server)

### Version

v1.5.9.2 (Summer Update Hotfix 2)

### Which operating system did you encounter this bug on?

Windows

**Comment by SomeRandomNoobKekeke** [comment]
Send save file

**Comment by Shadowdash19** [comment]
Dont sure where in post i should insert it so i just send it in comments
[Multiplayer.zip](https://github.com/user-attachments/files/17117141/Multiplayer.zip)

**Comment by SomeRandomNoobKekeke** [comment]
You got tons of laggy mods
![Screenshot_7](https://github.com/user-attachments/assets/dbe51d00-b12c-47f1-9b0c-8b707fcfb9bc)
And a super laggy Sub
![Screenshot_5](https://github.com/user-attachments/assets/d4d156fc-b816-4898-b4e4-3ac503ff04fc)
![Screenshot_6](https://github.com/user-attachments/assets/882ea3db-2cc4-43c2-baf6-7b9a5d663126)
> and after starting save without mods framerate was just slitly higher

Because most lags are comming from submarine

> our crew framerate really depends on items stashed on sub

Most vanilla items don't cause any lags, but modded items with status effects do

> we did arrive at 2nd location 

If performance dropped when you entered 2nd biome then it's probably caused by biome itself and it's wonderful water particles, background chunks and currents

### So
Try this https://steamcommunity.com/sharedfiles/filedetails/?id=2954718814

If your computer still can't handle it, switch to chad vanilla Dugong
![Screenshot_8](https://github.com/user-attachments/assets/e3e98b8d-abd6-482e-b229-feaa272c758f)

If it's still too laggy remove mods, starting with neurotrauma

**Reply by Shadowdash19** [reply]
I did mention that i tested out save with turned off mods and FPS just get higher on like 20%, not really much + we played with that sub (and mods turned on) earlier and had no troubles with performance

**Reply by SomeRandomNoobKekeke** [reply]
Yes, and in second biome lags from your sub and water particles stack and dropped your fps below 60

**Reply by Shadowdash19** [reply]
Removing particles seems to decrease my FPS ingame, maybe im wrong, but it definetly do NOT increase it in any seemable value, so particles is not the issue what killing my performance

**Reply by Shadowdash19** [reply]
also, i sent wrong save... that is actual one:
[Multiplayer.zip](https://github.com/user-attachments/files/17119243/Multiplayer.zip)

**Reply by Regalis11** [reply]
> it's probably caused by biome itself and it's wonderful water particles, background chunks and currents

This popular theory of why the biome is heavier than the others is actually completely wrong, the water particles have absolutely zero effect on performance compared to the other biomes (it's just a different texture, there's no difference or anything heavier about it), and the background chunks and currents have very little effect on performance. There is no need to guess, we can easily see what the issue is with a profiler (and it mainly has to do with the lighting on the level objects).

**Comment by SomeRandomNoobKekeke** [comment]
This sub is even laggier. Without items it's 2.5 times lagier than Berilia or Kastrull and 5 times laggier than Dugong
![20240924222210_1](https://github.com/user-attachments/assets/b2b159a9-d01f-4684-a976-c1c360c094d0)
![20240924222317_1](https://github.com/user-attachments/assets/0fe85caf-3dde-4433-9c03-a2760874aef9)
![20240924222407_1](https://github.com/user-attachments/assets/8ec9fa79-c0e3-4a54-b732-2b786a38d0b4)
I've no idea how you tested that without mods. I loaded that save without mods and there was 1 fps, all the items was lying on the floor, apparently because they was in some modded containers

All my fps was consumed by calculating physics of their bodies

If you have stuff just lying on the floor in your modded game, just put it all in containers and their physics will be disabled

**Comment by Regalis11** [comment]
Thank you for the report!

It is a known issue that large numbers of items can cause performance problems, and something we've done improvements to in almost every update (there's still more work to do to get it to perform even better). The reason why e.g. games like Minecraft has no problems with having a million items in your inventory or a chest is because those items don't need to do anything or really even "exist" - they can exist as just numbers. In Barotrauma, many contained items can do things: affect things around them or be affected by things happening around them. That's why contained items sometimes need to be updated in one way or another. 

In the vanilla game though, this has been taken into account, and most items that you can store large quantities of don't do anything, and don't need to be updated when in an inventory. Mods often don't take this (or performance considerations in general) into account, and the items might for example have several status effects that run every frame, which will lead to performance problems when there's many of those around.

It seems both of your saves are from a heavily modded game, and use very poorly optimized custom subs, so I'm afraid there's not much we can do to fix the performance in those by modifying the vanilla game.

Even though I think this report is somewhat valid and agree there's still a lot more optimization work to be done, I think I'll close the ticket: we also have the [General performance problems](https://github.com/FakeFishGames/Barotrauma/issues/9396) where this sort of information about performance issues can be posted.
